

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>stcrpy.tcr_processing.annotate &mdash; STCRpy 0.1.28 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=34399df3"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            STCRpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">stcrpy</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">STCRpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">stcrpy.tcr_processing.annotate</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for stcrpy.tcr_processing.annotate</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Created on 10 May 2017</span>
<span class="sd">@author: leem</span>

<span class="sd">Implementation to call anarci (built-in to STrDab) to annotate structures.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Polypeptide</span><span class="w"> </span><span class="kn">import</span> <span class="n">aa1</span><span class="p">,</span> <span class="n">aa3</span>  <span class="c1"># to allow me to return &quot;X&quot; if not found.</span>

<span class="n">to_one_letter_code</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">aa3</span><span class="p">,</span> <span class="n">aa1</span><span class="p">)))</span>

<span class="c1"># Import TCRDB&#39;s constants and common functions.</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.utils.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">TCR_CHAINS</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">anarci</span><span class="w"> </span><span class="kn">import</span> <span class="n">number</span> <span class="k">as</span> <span class="n">anarci_number</span>


<div class="viewcode-block" id="call_anarci">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.call_anarci">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">call_anarci</span><span class="p">(</span>
    <span class="n">seq</span><span class="p">,</span>
    <span class="n">allow</span><span class="o">=</span><span class="nb">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="s2">&quot;B&quot;</span><span class="p">,</span>
            <span class="s2">&quot;A&quot;</span><span class="p">,</span>
            <span class="s2">&quot;D&quot;</span><span class="p">,</span>
            <span class="s2">&quot;G&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GA1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GA2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GA1L&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GA2L&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GA&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GB&quot;</span><span class="p">,</span>
            <span class="s2">&quot;B2M&quot;</span><span class="p">,</span>
            <span class="s2">&quot;MH1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;MR1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;MR2&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">),</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Use the ANARCI program to number the sequence.</span>
<span class="sd">    @param seq: An amino acid sequence that you wish to number.</span>
<span class="sd">    @type seq: C{str}</span>

<span class="sd">    @return: numbering, chain type</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">numbering</span><span class="p">,</span> <span class="n">chain_type</span><span class="p">,</span> <span class="n">germline_info</span> <span class="o">=</span> <span class="n">anarci_number</span><span class="p">(</span>
        <span class="n">seq</span><span class="p">,</span> <span class="n">allow</span><span class="o">=</span><span class="n">allow</span><span class="p">,</span> <span class="n">assign_germline</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">numbering</span> <span class="ow">and</span> <span class="s2">&quot;MR&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">chain_type</span> <span class="ow">and</span> <span class="n">chain_type</span> <span class="ow">in</span> <span class="n">allow</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[(</span><span class="n">_</span><span class="p">,</span> <span class="n">aa</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">numbering</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">],</span> <span class="n">chain_type</span><span class="p">,</span> <span class="n">germline_info</span>
    <span class="k">elif</span> <span class="n">numbering</span> <span class="ow">and</span> <span class="n">chain_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;BA&quot;</span><span class="p">,</span> <span class="s2">&quot;GD&quot;</span><span class="p">,</span> <span class="s2">&quot;AB&quot;</span><span class="p">,</span> <span class="s2">&quot;DG&quot;</span><span class="p">]:</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="p">[[(</span><span class="n">_</span><span class="p">,</span> <span class="n">aa</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">aa</span> <span class="ow">in</span> <span class="n">n</span> <span class="k">if</span> <span class="n">aa</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">numbering</span><span class="p">],</span>
            <span class="n">chain_type</span><span class="p">,</span>
            <span class="n">germline_info</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="annotate">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.annotate">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">annotate</span><span class="p">(</span><span class="n">chain</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Annotate the sequence of a chain object from TCRDB.TcrPDB</span>
<span class="sd">    # e.g. if you have chains B, A and X, you want to force the annotator to return the annotation</span>
<span class="sd">    # for B and A but not for X (the antigen)</span>

<span class="sd">    returns a dictionary which has the residue ids as key and the annotation as value or is False,</span>
<span class="sd">    and chain type which is B/A/G/D/MH1/GA/GB/B2M or False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span> <span class="o">=</span> <span class="n">extract_sequence</span><span class="p">(</span><span class="n">chain</span><span class="p">)</span>
    <span class="n">numbering</span><span class="p">,</span> <span class="n">chain_type</span><span class="p">,</span> <span class="n">germline_info</span> <span class="o">=</span> <span class="n">call_anarci</span><span class="p">(</span><span class="n">sequence_str</span><span class="p">)</span>

    <span class="c1"># Use</span>
    <span class="k">if</span> <span class="n">chain_type</span><span class="p">:</span>
        <span class="n">chtype</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">chain_type</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">chtype</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">chtype</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;BA&quot;</span><span class="p">,</span> <span class="s2">&quot;GD&quot;</span><span class="p">):</span>
        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="n">align_scTCR_numbering</span><span class="p">(</span>
            <span class="n">numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span>
        <span class="p">)</span>
        <span class="c1"># aligned_numbering = cleanup_scTCR_numbering(aligned_numbering, sequence_list)</span>
        <span class="n">scTCR</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">elif</span> <span class="n">chtype</span> <span class="o">==</span> <span class="s2">&quot;DC1&quot;</span> <span class="ow">or</span> <span class="n">chtype</span> <span class="o">==</span> <span class="s2">&quot;RM1&quot;</span><span class="p">:</span>
        <span class="c1"># Use the scTCR numbering trick; since CD1/MR1 numbering only spans up to residue ~87 and</span>
        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="n">align_scTCR_numbering</span><span class="p">(</span>
            <span class="n">numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span>
        <span class="p">)</span>
        <span class="n">aligned_numbering</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">aligned_numbering</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="n">aligned_numbering</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># combine the numbering</span>
        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="n">cleanup_scTCR_numbering</span><span class="p">(</span><span class="n">aligned_numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">)</span>
        <span class="n">scTCR</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># align the original residue id&#39;s to the numbering</span>
        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="n">align_numbering</span><span class="p">(</span><span class="n">numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">)</span>
        <span class="n">scTCR</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># aligned numbering is a dictionary of the original residue ids and the new numbering</span>
    <span class="k">return</span> <span class="n">aligned_numbering</span><span class="p">,</span> <span class="n">chain_type</span><span class="p">,</span> <span class="n">germline_info</span><span class="p">,</span> <span class="n">scTCR</span></div>



<div class="viewcode-block" id="extract_sequence">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.extract_sequence">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_sequence</span><span class="p">(</span>
    <span class="n">chain</span><span class="p">,</span> <span class="n">selection</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">return_warnings</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ignore_hets</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">backbone</span><span class="o">=</span><span class="kc">False</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the amino acid sequence of the chain.</span>
<span class="sd">    @change:    Residues containing HETATOMs are skipped --&gt;  Residues containing HETATOMs are checked as an amino acid.</span>

<span class="sd">    Residues containing HETATOMs are checked  to be amino acids and the single letter returned.</span>

<span class="sd">    This works provided the residues in the chain are in the correct order.</span>

<span class="sd">    @param selection: a selection object to select certain residues</span>
<span class="sd">    @param return_warnings: Flag to return a list of warnings or not</span>
<span class="sd">    @param backbone: Flag whether to only show residues with a complete backbone (in the structure) or not.</span>
<span class="sd">    @return: The sequence in a resid:aa tuple list and the sequence as a string.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequence_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">warnings</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="o">.</span><span class="n">get_list</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot; &quot;</span>
        <span class="p">):</span>  <span class="c1"># skip HETATOMs - this is not necesserily a good idea, flag to the user that is has been done.</span>
            <span class="c1">#            if residue.get_resname() not in to_one_letter_code: # Check that the residue can be converted into a single letter.</span>
            <span class="c1">#                continue</span>
            <span class="c1">#            if residue.get_resname() in to_one_letter_code: # Check that the residue can be converted into a single letter.</span>
            <span class="c1">#                pass</span>
            <span class="k">if</span> <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">()</span> <span class="ow">in</span> <span class="n">to_one_letter_code</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ignore_hets</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">return_warnings</span><span class="p">:</span>
                        <span class="n">warnings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
<span class="w">                            </span><span class="sd">&quot;&quot;&quot;Warning: HETATM residue %s at position %s (PDB numbering) found in chain %s.</span>
<span class="sd">                            Not including it in structure&#39;s sequence.&quot;&quot;&quot;</span>
                            <span class="o">%</span> <span class="p">(</span>
                                <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">(),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                                <span class="n">residue</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
<span class="w">                            </span><span class="sd">&quot;&quot;&quot;Warning: HETATM residue %s position %s (PDB numbering) found in chain %s.</span>
<span class="sd">                            Not including it in structure&#39;s sequence.\n&quot;&quot;&quot;</span>
                            <span class="o">%</span> <span class="p">(</span>
                                <span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">(),</span>
                                <span class="nb">str</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                                <span class="n">residue</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>

        <span class="k">if</span> <span class="n">selection</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">selection</span><span class="o">.</span><span class="n">accept</span><span class="p">(</span><span class="n">residue</span><span class="p">):</span>
                <span class="k">continue</span>

        <span class="n">atoms_of_residue</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">child_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">backboneCondition</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;N&quot;</span> <span class="ow">in</span> <span class="n">atoms_of_residue</span>
            <span class="ow">and</span> <span class="s2">&quot;C&quot;</span> <span class="ow">in</span> <span class="n">atoms_of_residue</span>
            <span class="ow">and</span> <span class="s2">&quot;CA&quot;</span> <span class="ow">in</span> <span class="n">atoms_of_residue</span>
            <span class="ow">and</span> <span class="s2">&quot;O&quot;</span> <span class="ow">in</span> <span class="n">atoms_of_residue</span>
        <span class="p">)</span>  <span class="c1"># Boolean to hold if residue has a full backbone</span>

        <span class="c1"># CASE 1: backbone = True, and residue has a full backbone; convert a.a into single letter</span>
        <span class="k">if</span> <span class="n">backbone</span> <span class="ow">and</span> <span class="n">backboneCondition</span><span class="p">:</span>
            <span class="n">sequence_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">to_one_letter_code</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">(),</span> <span class="s2">&quot;X&quot;</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="c1"># CASE 2: backbone = True, but residue does not have a full backbone; use a gap in sequence annotation</span>
        <span class="k">elif</span> <span class="n">backbone</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">backboneCondition</span><span class="p">:</span>
            <span class="n">sequence_list</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">))</span>
        <span class="c1"># CASE 0 (default): don&#39;t care about backbone, just write it to sequence if it&#39;s found in structure.</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">backbone</span><span class="p">:</span>
            <span class="n">sequence_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">to_one_letter_code</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">get_resname</span><span class="p">(),</span> <span class="s2">&quot;X&quot;</span><span class="p">))</span>
            <span class="p">)</span>  <span class="c1"># i am</span>

    <span class="n">sequence_str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">sequence_list</span><span class="p">])</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">return_warnings</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span><span class="p">,</span> <span class="n">warnings</span></div>



<div class="viewcode-block" id="interpret">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.interpret">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">interpret</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to interpret an annotation in the form H100A into the form ( 100, &#39;A&#39; )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">TCR_CHAINS</span><span class="p">,</span> <span class="n">x</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]),</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]),</span> <span class="n">x</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span></div>



<div class="viewcode-block" id="align_numbering">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.align_numbering">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">align_numbering</span><span class="p">(</span><span class="n">numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">alignment_dict</span><span class="o">=</span><span class="p">{}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Align the sequence that has been numbered to the sequence you input.</span>
<span class="sd">    The numbered sequence should be &quot;in&quot; the input sequence.</span>
<span class="sd">    If not, supply an alignment dictionary.(align sequences and use get_alignment_dict(ali1,ali2))</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">numbering</span><span class="p">:</span>
        <span class="n">numbered_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">numbering</span><span class="p">])</span>
        <span class="n">input_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">sequence_list</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">alignment_dict</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">numbered_sequence_ali</span><span class="p">,</span> <span class="n">input_sequence_ali</span> <span class="o">=</span> <span class="n">pairwise_alignment</span><span class="p">(</span>
                    <span class="n">numbered_sequence</span><span class="p">,</span> <span class="n">input_sequence</span>
                <span class="p">)</span>
                <span class="n">alignment_dict</span> <span class="o">=</span> <span class="n">get_alignment_dict</span><span class="p">(</span>
                    <span class="n">input_sequence_ali</span><span class="p">,</span> <span class="n">numbered_sequence_ali</span>
                <span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                    <span class="s2">&quot;Could not align numbered sequence to aligned sequence:&quot;</span>
                    <span class="o">+</span> <span class="s2">&quot; &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">numbered_sequence</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">input_sequence</span><span class="p">)</span>
                <span class="p">)</span>

        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">n</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="n">after_flag</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">input_sequence</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">alignment_dict</span><span class="p">:</span>
                <span class="c1"># during</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">after_flag</span> <span class="ow">is</span> <span class="kc">False</span>
                <span class="p">),</span> <span class="s2">&quot;Extra residue in structure than expected from provided sequence&quot;</span>
                <span class="k">assert</span> <span class="p">(</span>
                    <span class="n">input_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">numbered_sequence</span><span class="p">[</span><span class="n">alignment_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
                <span class="p">),</span> <span class="s2">&quot;alignment dictionary failed&quot;</span>
                <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">numbering</span><span class="p">[</span><span class="n">alignment_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">n</span> <span class="o">=</span> <span class="n">numbering</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="c1"># after</span>
                <span class="n">after_flag</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
                <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># before numbering</span>
                <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">return</span> <span class="n">aligned_numbering</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="align_scTCR_numbering">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.align_scTCR_numbering">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">align_scTCR_numbering</span><span class="p">(</span><span class="n">numbering</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">,</span> <span class="n">sequence_str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Align the sequence that has been numbered to a scTCR structure.</span>
<span class="sd">    @param numbering:     numbered list of residues; this is usually a two-element list/tuple from TCRDB.anarci.number</span>
<span class="sd">    @param sequence_list: list of residues (e.g. from a structure) in its original numbering</span>
<span class="sd">    @param sequence_str:  string form of sequence_list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">numbering</span><span class="p">:</span>
        <span class="n">numbered_sequence</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">n</span><span class="p">])</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">numbering</span><span class="p">]</span>
        <span class="n">input_sequence</span> <span class="o">=</span> <span class="n">sequence_str</span>

        <span class="n">aligned_numbering</span> <span class="o">=</span> <span class="p">[{},</span> <span class="p">{}]</span>

        <span class="k">for</span> <span class="n">ii</span><span class="p">,</span> <span class="n">a_sequence</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">numbered_sequence</span><span class="p">):</span>

            <span class="c1"># Align each of the joined sequences from the numbering into the target structure sequence in &quot;sequence_str&quot;</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">a_sequence_ali</span><span class="p">,</span> <span class="n">input_sequence_ali</span> <span class="o">=</span> <span class="n">pairwise_alignment</span><span class="p">(</span>
                    <span class="n">a_sequence</span><span class="p">,</span> <span class="n">input_sequence</span>
                <span class="p">)</span>
                <span class="n">alignment_dict</span> <span class="o">=</span> <span class="n">get_alignment_dict</span><span class="p">(</span><span class="n">input_sequence_ali</span><span class="p">,</span> <span class="n">a_sequence_ali</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span>
                    <span class="s2">&quot;Could not align numbered sequence to aligned sequence&quot;</span>
                    <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">numbered_sequence</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">input_sequence</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="n">n</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="n">after_flag</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="c1"># for i in xrange(len(input_sequence)):</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">alignment_dict</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">alignment_dict</span><span class="p">:</span>
                    <span class="c1"># during</span>
                    <span class="k">assert</span> <span class="p">(</span>
                        <span class="n">after_flag</span> <span class="ow">is</span> <span class="kc">False</span>
                    <span class="p">),</span> <span class="s2">&quot;Extra residue in structure than expected from provided sequence&quot;</span>
                    <span class="k">assert</span> <span class="p">(</span>
                        <span class="n">input_sequence</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">numbered_sequence</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">alignment_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
                    <span class="p">),</span> <span class="s2">&quot;alignment dictionary failed&quot;</span>
                    <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">numbering</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span>
                        <span class="n">alignment_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">n</span> <span class="o">=</span> <span class="n">numbering</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># after</span>
                    <span class="n">after_flag</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
                    <span class="n">n</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># before numbering</span>
                    <span class="n">aligned_numbering</span><span class="p">[</span><span class="n">ii</span><span class="p">][</span><span class="n">sequence_list</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">return</span> <span class="n">aligned_numbering</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="cleanup_scTCR_numbering">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.cleanup_scTCR_numbering">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">cleanup_scTCR_numbering</span><span class="p">(</span><span class="n">numbering_dict</span><span class="p">,</span> <span class="n">sequence_list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The scTCR numbering method, while useful for sequences with two domains,</span>
<span class="sd">    can have gaps in between (e.g. CD1 molecule of 4lhu).</span>
<span class="sd">    This is to close the gaps in the numbering so that residues that were unnumbered by anarci don&#39;t move around</span>
<span class="sd">    during structural parsing (when they&#39;re probably just connections between domains).</span>

<span class="sd">    @param numbering_dict: numbered dictionary from align_scTCR_numbering</span>
<span class="sd">    @param sequence_list : sequence list from the structure for alignment.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">sequence_list</span><span class="p">]</span>

    <span class="c1"># This gets the last numbered residue in numbering_dict</span>
    <span class="n">lastkey</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">numbering_dict</span><span class="p">)</span>
    <span class="n">lastidx</span> <span class="o">=</span> <span class="n">positions</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">lastkey</span><span class="p">)</span>  <span class="c1"># Where is this on sequence_list?</span>

    <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)):</span>

        <span class="c1"># If we got to the last key, don&#39;t bother.</span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">&gt;</span> <span class="n">lastidx</span><span class="p">:</span>
            <span class="k">break</span>

        <span class="n">key</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>

        <span class="c1"># If a target key is not in the numbering dict, see where it fits, then fit a number in it.</span>
        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">numbering_dict</span><span class="p">:</span>

            <span class="c1"># Get the left and right bounds of the gap</span>
            <span class="n">left</span><span class="p">,</span> <span class="n">right</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span>
            <span class="n">lidx</span><span class="p">,</span> <span class="n">ridx</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
            <span class="n">lval</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="c1"># Continue iterating left from the missing key until we find one that exists</span>
            <span class="k">while</span> <span class="ow">not</span> <span class="n">left</span><span class="p">:</span>
                <span class="n">key_left</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">index</span> <span class="o">-</span> <span class="n">j</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">key_left</span> <span class="ow">in</span> <span class="n">numbering_dict</span><span class="p">:</span>
                    <span class="n">left</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">lidx</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">index</span> <span class="o">-</span> <span class="n">j</span>
                    <span class="p">)</span>  <span class="c1"># Last known index of sequence_list where we know a key exists</span>
                    <span class="n">lval</span> <span class="o">=</span> <span class="n">numbering_dict</span><span class="p">[</span><span class="n">key_left</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">while</span> <span class="ow">not</span> <span class="n">right</span><span class="p">:</span>
                <span class="n">key_right</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[</span><span class="n">index</span> <span class="o">+</span> <span class="n">j</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">key_right</span> <span class="ow">in</span> <span class="n">numbering_dict</span><span class="p">:</span>
                    <span class="n">right</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">ridx</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">index</span> <span class="o">+</span> <span class="n">j</span>
                    <span class="p">)</span>  <span class="c1"># Last known index of sequence_list on the right where we know a key exists</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="c1"># For every key between the left and right, fill in</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">missing_key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="n">lidx</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:</span> <span class="n">ridx</span><span class="p">]):</span>
                <span class="n">numbering_dict</span><span class="p">[</span><span class="n">missing_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">lval</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">numbering_dict</span></div>



<div class="viewcode-block" id="get_alignment_dict">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.get_alignment_dict">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_alignment_dict</span><span class="p">(</span><span class="n">ali1</span><span class="p">,</span> <span class="n">ali2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get a dictionary which tells you the index in sequence 2 that should align with the index in sequence 1 (key)</span>

<span class="sd">    ali1:  ----bcde-f---        seq1: bcdef</span>
<span class="sd">    ali2:  ---abcd--f---        seq2: abcdf</span>

<span class="sd">    alignment_dict={</span>
<span class="sd">        0:1,</span>
<span class="sd">        1:2,</span>
<span class="sd">        2:3,</span>
<span class="sd">        4:4</span>
<span class="sd">        }</span>

<span class="sd">    If the index is aligned with a gap do not include in the dictionary.</span>
<span class="sd">    e.g  1 in alignment_dict  --&gt; True</span>
<span class="sd">    e.g  3 in alignment_dict  --&gt; False</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">ali1</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">ali2</span>
    <span class="p">),</span> <span class="s2">&quot;aligned sequences must be same lengths (including gaps)&quot;</span>
    <span class="n">alignment_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">p1</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">p2</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="k">for</span> <span class="n">ap</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ali1</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">ali1</span><span class="p">[</span><span class="n">ap</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="ow">and</span> <span class="n">ali2</span><span class="p">[</span><span class="n">ap</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
            <span class="n">p1</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">p2</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">alignment_dict</span><span class="p">[</span><span class="n">p1</span><span class="p">]</span> <span class="o">=</span> <span class="n">p2</span>
        <span class="k">elif</span> <span class="n">ali1</span><span class="p">[</span><span class="n">ap</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
            <span class="n">p1</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">ali2</span><span class="p">[</span><span class="n">ap</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
            <span class="n">p2</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">alignment_dict</span></div>



<div class="viewcode-block" id="pairwise_alignment">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.pairwise_alignment">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">pairwise_alignment</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">exact</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to do alignment of sequences between sequences using biopython.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="n">warnings</span><span class="o">.</span><span class="n">catch_warnings</span><span class="p">():</span>  <span class="c1"># prevents pairwise2 deprecation warning from being raised</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">Bio.pairwise2</span><span class="w"> </span><span class="kn">import</span> <span class="n">align</span>

    <span class="n">alignment</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">s1_aln</span><span class="p">,</span> <span class="n">s2_aln</span> <span class="o">=</span> <span class="n">easy_alignment</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">s1_aln</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">s1_aln</span><span class="p">,</span> <span class="n">s2_aln</span>

    <span class="k">if</span> <span class="n">exact</span><span class="p">:</span>
        <span class="c1"># Align with a match score of 1, mismatch of 0, gap opening of -1.001, and gap extension of -1</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">globalms</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">1.001</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">alignment</span><span class="p">:</span>
        <span class="n">aligned_seqs</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">aligned_seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">aligned_seqs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="easy_alignment">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.easy_alignment">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">easy_alignment</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to align two sequences by checking if one is in the other.</span>
<span class="sd">    This function will conserve gaps.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="p">(</span>
        <span class="nb">type</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">str</span> <span class="ow">and</span> <span class="nb">type</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">str</span>
    <span class="p">),</span> <span class="s2">&quot;Sequences must be strings for easy_alignment&quot;</span>
    <span class="k">if</span> <span class="n">seq1</span> <span class="ow">in</span> <span class="n">seq2</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">seq2</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
        <span class="n">seq1_ali</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="n">start</span> <span class="o">+</span> <span class="n">seq1</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span> <span class="o">-</span> <span class="n">start</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">seq1_ali</span><span class="p">,</span> <span class="n">seq2</span>

    <span class="k">elif</span> <span class="n">seq2</span> <span class="ow">in</span> <span class="n">seq1</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">seq1</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
        <span class="n">seq2_ali</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="n">start</span> <span class="o">+</span> <span class="n">seq2</span> <span class="o">+</span> <span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">-</span> <span class="n">start</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2_ali</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Can&#39;t align them # I return just one value here.</span>
        <span class="k">return</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span></div>



<div class="viewcode-block" id="validate_sequence">
<a class="viewcode-back" href="../../../stcrpy.tcr_processing.html#stcrpy.tcr_processing.annotate.validate_sequence">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">validate_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check whether a sequence is a protein sequence or if someone has submitted something nasty.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">10000</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span><span class="s2">&quot;Sequence too long.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">([</span><span class="mi">1</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">if</span> <span class="n">s</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">aa1</span><span class="p">]):</span>
        <span class="k">raise</span> <span class="ne">AssertionError</span><span class="p">(</span>
            <span class="s2">&quot;Unknown amino acid letter found in sequence: &quot;</span> <span class="o">+</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="k">pass</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Nele P. Quast.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>